druid-shell/src/scale.rs

```diff
@@ -0,0 +1,388 @@
+/// often not limited to integers. This allows for accurate calculations of
+/// the platform area pixel boundaries from the logcal area using a [`Scale`].
+///
+/// Even though the logcal area size can be fractional, the integer boundaries of that logical area
+/// will still match up with the platform area pixel boundaries as often as the scale factor allows.
+///
+/// A copy of `ScaledArea` will be stale as soon as the platform area size changes.
+///
+/// [`Scale`]: struct.Scale.html
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub struct ScaledArea {
+    /// The size of the scaled area in display points.
+    size_dp: Size,
+    /// The size of the scaled area in pixels.
+    size_px: Size,
+}
+
+/// The `Scalable` trait describes how coordinates should be translated
+/// from display points into pixels and vice versa using a [`Scale`].
+///
+/// [`Scale`]: struct.Scale.html
+pub trait Scalable {
+    /// Converts the scalable item from display points into pixels,
+    /// using the x axis scale factor for coordinates on the x axis
+    /// and the y axis scale factor for coordinates on the y axis.
+    fn to_px(&self, scale: &Scale) -> Self;
+
+    /// Converts the scalable item from pixels into display points,
+    /// using the x axis scale factor for coordinates on the x axis
+    /// and the y axis scale factor for coordinates on the y axis.
+    fn to_dp(&self, scale: &Scale) -> Self;
+}
+
+impl Default for Scale {
+    fn default() -> Scale {
+        Scale {
+            dpi_x: SCALE_TARGET_DPI,
+            dpi_y: SCALE_TARGET_DPI,
+            scale_x: 1.0,
+            scale_y: 1.0,
+        }
+    }
+}
+
+impl Scale {
+    /// Create a new `Scale` state based on the specified DPIs.
+    ///
+    /// Use this constructor if the platform provided DPI is the most accurate number.
+    pub fn from_dpi(dpi_x: f64, dpi_y: f64) -> Scale {
+        Scale {
+            dpi_x,
+            dpi_y,
+            scale_x: dpi_x / SCALE_TARGET_DPI,
+            scale_y: dpi_y / SCALE_TARGET_DPI,
+        }
+    }
+
+    /// Create a new `Scale` state based on the specified scale factors.
+    ///
+    /// Use this constructor if the platform provided scale factor is the most accurate number.
+    pub fn from_scale(scale_x: f64, scale_y: f64) -> Scale {
+        Scale {
+            dpi_x: SCALE_TARGET_DPI * scale_x,
+            dpi_y: SCALE_TARGET_DPI * scale_y,
+            scale_x,
+            scale_y,
+        }
+    }
+
+    /// Returns the x axis platform DPI associated with this `Scale`.
+    #[inline]
+    pub fn dpi_x(&self) -> f64 {
+        self.dpi_x
+    }
+
+    /// Returns the y axis platform DPI associated with this `Scale`.
+    #[inline]
+    pub fn dpi_y(&self) -> f64 {
+        self.dpi_y
+    }
+
+    /// Returns `true` if the specified DPI is approximately equal to the `Scale` DPI.
+    #[inline]
+    pub fn dpi_approx_eq(&self, dpi_x: f64, dpi_y: f64) -> bool {
```

Comment:
is there a specific rationale for this, instead of say making the caller create a `Scale` and having an `approx_eq` for that?

Reply:
The rationale has vanished. This used to be useful with an initial implementation that used a different strategy and attempted to always achieve integer logical display points. That no longer being the case combined with `Scale` now just being the scale factors, we can remove this. The new `Scale` implements `PartialEq` which will be just fine for our needs.